zaza/charm_tests/keystone/setup.py

```diff
@@ -0,0 +1,116 @@
+    def _v2():
+        keystone_session = openstack_utils.get_overcloud_keystone_session()
+        keystone_client = openstack_utils.get_keystone_session_client(
+            keystone_session, api_version=2)
+        tenant = keystone_client.tenants.create(tenant_name=DEMO_TENANT,
+                                                description='Demo Tenant',
+                                                enabled=True)
+        keystone_client.users.create(name=DEMO_USER,
+                                     password=DEMO_PASSWORD,
+                                     tenant_id=tenant.id)
+
+    def _v3():
+        keystone_session = openstack_utils.get_overcloud_keystone_session()
+        keystone_client = openstack_utils.get_keystone_session_client(
+            keystone_session)
+        domain = keystone_client.domains.create(
+            DEMO_DOMAIN,
+            description='Demo Domain',
+            enabled=True)
+        project = keystone_client.projects.create(
+            DEMO_PROJECT,
+            domain,
+            description='Demo Project',
+            enabled=True)
+        demo_user = keystone_client.users.create(
+            DEMO_USER,
+            domain=domain,
+            project=project,
+            password=DEMO_PASSWORD,
+            email='demo@demo.com',
+            description='Demo User',
+            enabled=True)
+        member_role = keystone_client.roles.find(name='Member')
+        keystone_client.roles.grant(
+            member_role,
+            user=demo_user,
+            project_domain=domain,
+            project=project)
+        demo_admin_user = keystone_client.users.create(
+            DEMO_ADMIN_USER,
+            domain=domain,
+            project=project,
+            password=DEMO_ADMIN_USER_PASSWORD,
+            email='demo_admin@demo.com',
+            description='Demo Admin User',
+            enabled=True)
+        admin_role = keystone_client.roles.find(name='Admin')
+        keystone_client.roles.grant(
+            admin_role,
+            user=demo_admin_user,
+            domain=domain)
+        keystone_client.roles.grant(
+            member_role,
+            user=demo_admin_user,
+            project_domain=domain,
+            project=project)
+        keystone_client.roles.grant(
+            admin_role,
+            user=demo_admin_user,
+            project_domain=domain,
+            project=project)
+
+    if (openstack_utils.get_os_release() <
+            openstack_utils.get_os_release('trusty_mitaka')):
+        # create only V2 user
+        _v2()
+        return
+
+    if (openstack_utils.get_os_release() >=
+        openstack_utils.get_os_release('trusty_mitaka') and
+        openstack_utils.get_os_release() <
+            openstack_utils.get_os_release('xenial_queens')):
+        # create V2 and V3 user
+        _v2()
+
+        _singleton = BaseKeystoneTest()
+        _singleton.setUpClass()
+        with _singleton.config_change(
```

Comment:
@gnuoy I would like your advice on this.

While this solves the problem at hand of using the context managed config_change function from a test class in my setup function, I get a feeling of doing mischief and fear I will be eaten by a rabbit (or something of the sort) around the next corner.

Would you think this is OK, or should I attempt to reorganize so I can more cleanly share code between my setup functions and test classes?

Any advice appreciated! :-)

Reply:
I understand your concern, it does feel like a module boundary violation. The only thing that occurred to me was to move config_change out of the OpenStackBaseTest class to be a standalone method in zaza.charm_tests.test_utils. We'd have to add application_name, model_name and states options to it. Then add back a config_change method to OpenStackBaseTest that calls the new method filling in the new options. However, I'm not convinced that that approach actually buys us much. TBH I'm happy to leave it as it is as I don't think this is going to be a common pattern.

Reply:
Thank you for your advice on this, much appreciated! :+1: 

I have since moved my base class out to `__init__.py` to avoid the circular import problem, but other than that I will leave it then.  Thx!